### ISSUE TITLE
Generate series of values that are dependent on the previous value

### ISSUE THREAD

**Issue by rynobax** [dubzzz/fast-check#424]
## 💬 Question and Help

I have a reducer that maintains a set of numbers stored as an array.  I'm trying to generate a series of actions to test the reducer with.  (This is a simplified example of the actual problem)

```
[1, 2]
```
And there are the following actions:
```
{ type: "add", value: number }
{ type: "remove", value: number }
```
I want to generate a series of actions to dispatch, where the actions are dependent on the state _after_ dispatching the previous action.  For example, two remove actions for the same value would be invalid, because the first action would have already removed the value.

Valid Examples:
```
initial state: [1, 2]
// Can remove either of the start values
[{ type: "remove", value: 1 }, { type: "remove", value: 2 }]
// even though 3 is not in the initial state, after it's added it becomes a valid remove target
[{ type: "add", value: 3 }, { type: "remove", value: 3 }]
```

Invalid Examples:
```
initial state: [1, 2]
// Should not be able to remove a value not in state
[{ type: "remove", value: 3 }]
// Should not be able to remove a value twice
[{ type: "remove", value: 2 }, { type: "remove", value: 2 }]
```

I can generate an Arbitrary for valid actions based on a state, I just can't figure out how to create an array where each element is dependent on the value of the previous.

I've poked around Model based testing / Commands, but it seems like it has the same problem, where you must define all your commands up front.

So my question is:  Is this use case something the library currently supports?  If not, is it something that could be added?  I'd be happy to help out if so!

**Comment by dubzzz** [comment]
Hi @rynobax,

Here is a suggestion of how I'd have tested it using fast-check today:

```js
const fc = require('fast-check');

// The system under test

const initialState = {
    data: []
};
const reducer = (state = initialState, action) => {
    switch (action.type) {
        case 'add': {
            return { ...state, data: state.data.concat([action.value]) };
        }
        case 'remove': {
            const index = state.data.indexOf(action.value);
            if (index === -1)
                throw new Error(`Unable to delete ${action.value} from state ${JSON.stringify(state)}`);
            
            //// Implementation: No bug
            // const newData = state.data.slice();
            // newData.splice(index, 1);
            
            // Implementation: With bug
            const newData = state.data.filter(v => v !== action.value);
            return { ...state, data: newData };
        }
        default: {
            throw new Error(`No such action ${action.type}`);
        }
    }
};

// The commands (model based testing approach)

class AddCommand {
    constructor(value) {
        this.value = value;
    }
    check(m) {
        // we can always add the value
        return true;
    }
    run(m, r) {
        m.push(this.value);
        r.state = reducer(r.state, { type: 'add', value: this.value });
    }
    toString() {
        return `new AddCommand(${this.value})`;
    }
}

class RemoveCommand {
    constructor(value) {
        this.value = value;
    }
    check(m) {
        // we can only remove the value if the value is there
        // thus we accept to run this command only if the value is within our set
        return m.includes(this.value);
    }
    run(m, r) {
        m.splice(m.indexOf(this.value), 1);
        r.state = reducer(r.state, { type: 'remove', value: this.value });
    }
    toString() {
        return `new RemoveCommand(${this.value})`;
    }
}

// The test

fc.assert(
    fc.property(
        fc.array(fc.hexa()),
        fc.commands([
            fc.hexa().map(v => new AddCommand(v)),
            fc.hexa().map(v => new RemoveCommand(v)),
        ]),
        (initialState, cmds) => {
            const s = () => ({
                model: initialState.slice(), // as we will mute it, we clone it
                real: { state: { data: initialState } }, // this one is never muted, we always create a new one in the reducer
            });
            fc.modelRun(s, cmds);
        }));
```
Try it online: https://runkit.com/dubzzz/5d718c582eab5b001379bffc

The implementation above is based on Model Based Testing approach which is documented at:
- https://github.com/dubzzz/fast-check/blob/master/documentation/1-Guides/Arbitraries.md#model-based-testing
- https://github.com/dubzzz/fast-check/blob/master/documentation/1-Guides/Tips.md#model-based-testing-or-ui-test

Another way to test it would have been to use the `chain` operator provided by the Arbitrary type to chain all arbitraries together. Maybe less adapted for this need.

Does it solves your problem and answer your needs?

**Comment by rynobax** [comment]
Thanks for the answer, I appreciate your help with this!

I should adjust my sample problem to better match what I am trying to do:

Instead of add having the signature `{ type: "add", value: number }`, it should be just `{ type: "add" }`.  The `value` for the add will be randomly generated inside of the reducer.  With the solution above, I would be unable to generate values for the remove actions that would match the random values that would end up in the state.

I think I could refactor the add action to have the ID passed in, and then I could get your example to work for me!  However I would be super interested in a solution that works with the updated problem.

**Comment by dubzzz** [comment]
I think that I'll need a more concrete description of the issue to help you a bit more on that point. Maybe a minimal reproduction of the reducer code.

Please note that generating random values within your reducer code (outside of fast-check control) might break the shrink and replay capabilities provided my fast-check (mostly if the randomly generated value affects the way your code behave). You can have a look into https://github.com/dubzzz/fast-check/blob/master/documentation/1-Guides/Tips.md#combine-with-other-faker-or-random-generator-libraries to see how you can combine fast-check with other (seeded) sources of random. Basically you can create a fake `Math.random` to pass to your code with (this one does not break shrinking and replay capabilities):

```js
const mathRandomArb = fc.infiniteStream(fc.double().noBias())
  .noShrink()
  .map(s => {
    const rng = () => s.next().value; // prng like Math.random but controlled by fast-check
    return rng;
  });
```

**Comment by rynobax** [comment]
So I think I finally got something working, by calling `.chain` in a loop.  It looks like this:

```
// editorReducer is my reducer function
// StateProducer is an Arbitrary<State>
// ActionProducer is a fn of shape (s: State) => Arbitrary<Action>, where the resulting
//   Arbitrary will only generate valid actions for that state

const JourneyProducer: fc.Arbitrary<[State, Action[]]> = fc
  // Create inital state and number of actions to generate
  .tuple(StateProducer, fc.integer(1, 10))
  .chain(([initialState, actionCount]) => {
    let builder = fc.tuple(
      fc.constant([] as Action[]),
      ActionProducer(initialState)
    );

    // Call chain for each action we want
    for (let i = 0; i < actionCount; i++) {
      builder = builder.chain(([actions, nextAction]) => {
        // Compute the state after all the actions so far
        const newActions = [...actions, nextAction];
        const latestState = newActions.reduce(editorReducer, initialState);

        // Return all the actions so far, and a arbitrary based on the latest state
        return fc.tuple(fc.constant(newActions), ActionProducer(latestState));
      });
    }

    // Return the initial state and list of actions
    return builder.map(([actions]) => [initialState, actions]);
  });
```

Does this seem like a reasonable approach?  It it doesn't I can put together a more complete example 😅

**Comment by dubzzz** [comment]
> Does this seem like a reasonable approach?

Actually it highly depends on what you want to test within your `fc.assert(fc.property(

**Comment by rynobax** [comment]
Alright, l can put together a working example of what I am trying to test and post it here in a few days.

Thanks again for the help!

**Comment by dubzzz** [comment]
@rynobax Any updates?

**Comment by rynobax** [comment]
[Here is a codesandbox with what I ended up with.](https://codesandbox.io/embed/awesome-chaum-ty79x?fontsize=14&previewwindow=tests)  I still feel it would be more intuitive to test a reducer with a series of actions where each action is based on the previous state, but using model based testing worked.  I'm not familiar with how shrinking works, so that may be impossible.

I am satisfied if you want to close the issue.  Also happy to answer questions if you have any.

**Comment by dubzzz** [comment]
Great! Your test in codesandbox looks how I suggested it.

> be more intuitive to test a reducer with a series of actions where each action is based on the previous state

@rynobax Let's go deeper and fully answer your problem.

I wanted to dig a little bit more in your issue. While `check` method and model based approach partially answers your needs I think we can go further.

Keep in mind that the generation phase is not supposed to contain any assertions or test your code. But it can use complex helpers to construct data that will be used by your system during the test.

By tweaking a bit the generators I got this stuff:

```
const fc = require('fast-check')

// Helpers

const text = id => ({ id, content: "", type: "Text" });

// Arbitraries

const allIds = ['1', '2', '3', '4', '5'];
const dirArb = fc.constantFrom("UP", "DOWN", "LEFT", "RIGHT");

const initArb = () => fc.record({
    type: fc.constant('INIT'),
    node: fc.constantFrom(...allIds).map(text)
});
const removeArb = (ids) => fc.record({
    type: fc.constant('REMOVE'),
    node: fc.constantFrom(...ids).map(text)
});
const addArb = (ids, destIds) => fc.record({
    type: fc.constant('ADD'),
    node: fc.constantFrom(...ids).map(text),
    destId: fc.constantFrom(...destIds),
    destDirection: dirArb
});
const moveArb = (ids) => fc.record({
    type: fc.constant('MOVE'),
    srcId: fc.constantFrom(...ids),
    desId: fc.constantFrom(...ids),
    destDirection: dirArb
});

const simpleModelReducer = (m, c) => {
    switch (c.type) {
        case 'INIT':
            return [...m, c.node.id];
        case 'REMOVE':
            return m.filter(it => it !== c.node.id);
        case 'ADD':
            return [...m, c.node.id];
        case 'MOVE':
            return m;
    }
};

const chaining = ({ m, cmds }) => {
    // We only want 10 commands in our scenario
    // We reached our limit ;)
    if (cmds.length === 10)
        return fc.constant({ m, cmds });
    
    // We need more commands for our scenario
    // We select all the valid generators
    const validCmds = [];
    const nonSelectedIds = allIds.filter(i => !m.includes(i));
    if (m.length === 0) validCmds.push(initArb());
    if (m.length > 0) validCmds.push(removeArb(m));
    if (nonSelectedIds.length > 0 && m.length > 0) validCmds.push(addArb(nonSelectedIds, m));
    if (m.length > 0) validCmds.push(moveArb(m));
    
    return longerSequenceArb = fc.oneof(...validCmds)
        .chain(c => {
            return chaining({ m: simpleModelReducer(m, c), cmds: [...cmds, c] });
        });
};
const cmdsRawArb = fc.constant({ m: [], cmds: [] }).chain(chaining);
const cmdsArb = cmdsRawArb.map(({ cmds }) => cmds);

fc.sample(cmdsArb, 1)
```

Basically `cmdsArb` is a generator of scenario as you wanted. If it answers your needs there is definitely something to add into fast-check to cover this kind of need ;)

Runkit link: https://runkit.com/dubzzz/5d9cf3598a931b001292be36

**Comment by dubzzz** [comment]
@rynobax Or an updated version of the snippet suggested above: https://runkit.com/dubzzz/fast-check-scenario-builder

The main issue I see for the moment, is that it will provide poor shrinking capabilities contrary to commands (but I'll check if I can come with something better for `scenarioArb` if it makes sense to be added into fast-check).

**Comment by rynobax** [comment]
That snippet is helpful, it looks like the kind of API I wanted

**Comment by nth-commit** [comment]
Huge +1 for something like this.

The model framework didn't really work for me either, as I needed my actions to be much more arbitrary. I also needed to specify a weighting on the action type, so that I can get the state to actually grow over time (I noticed you had the same issue above, most of the time the state would be empty, as the clear action undid everything).

Also, in my use-case, I didn't find it useful to track a simplified model in the test code. The test code uses the real reducer to build up the state. Then, I will have assertions against the state e.g. "The value of X can never be negative". I will also have assertions after a single action is dispatched on an arbitrary state, e.g. "Given any state, when I increment, the value of X' = X+1". The benefit of this approach is you can guarantee rules of your state are never violated, no matter what action(s) came before (i.e. "properties").

**Comment by nth-commit** [comment]
The other thing that's worth calling out, is because we want to guarantee our properties hold true forall scenarios, then it's not just the final state we want to test, it's all the intermediate states as well. That was the motivation for this issue: https://github.com/dubzzz/fast-check/issues/638

**Comment by nth-commit** [comment]
Hey @dubzzz, in thinking more about this, I think it would be pretty succinctly captured by having an `unfold` function.

```typescript
const unfold = <T>(seed: T, unfolder: (previous: T) => Arbitrary<T | undefined>): Arbitrary<T> => ...
```

You could leverage a function like that to easily test a "scenario" in a pretty flexible way, e.g.:

```typescript
const scenario = <State>(initialState: Arbitrary<State>, getNextState: (s: State) => Arbitrary<State | undefined>): Arbitrary<State> => ...
```

Usage:

```typescript
type Action = ...;

type State = ...;

const initialStateArbitrary = (): fc.Arbitrary<State> => { };

const actionArbitrary = (state: State): fc.Arbitrary<Action | undefined> => {
  // This function can decide what the actions are valid against the current state
};

const stateArbitrary = (reducer: (s: State, a: Action) => State): fc.Arbitrary<State> =>
  initialStateArbitrary().chain(initialState => {
    return fc.scenario(
      initialState,
      s => fc.frequency(
        {
          weight: 1, // Sometimes I want to terminate and start fresh
          arbitrary: fc.constant(undefined)
        },
        {
          weight: 20, // But mostly I want to apply an arbitrary action to the last state
          arbitrary: actionArbitrary(s).map(a => reducer(s, a))
        }
      });
  });
```

**Comment by dubzzz** [comment]
Yes, that might be one of the possibilities. By the way, it is quite close to the current proposal in https://github.com/dubzzz/fast-check/pull/445 except that selecting the action is based on an arbitrary.

The scenario approach has a major drawback compared to what is offered by the model approach of today: it is hard to shrink. I think it is one of the major difficulties with scenarios: how to make them shrinkable? If not feasible, then it should not be part of the framework as the framework will not offer any advantages (current draft PR could simply be an arbitrary manually written or offered by another lib if needed).

In the current model based testing approach, the framework generates long arrays of commands such a [ca, cb, cc, cd, ce...]. But when running the commands, if for instance cc fails and cb was not executed (check failed), it will shrink [ca, cc] and not the whole array.

**Comment by l-abels** [comment]
I want to start using `fast-check`, but I think the system I want to test might have the same issues described here.  I have a simple nested data structure that looks like this:
```
Foo {
   children: Bar[]
}

Bar {
    children: Baz[]
}

Baz {
    value: Int
}
```

Using `fc.commands`, I see that it would be pretty simple to test a command like `AddBarToFoo`.  Commands like `DeleteBar` and `AddBazToBar` would obviously need an index or id for a `Bar` to work, which depends on the state.  Am I correct in assuming that this system needs `fc.scenario` to be tested properly?

**Comment by nth-commit** [comment]
Yep looks that way to me, @l-abels.

**Comment by dubzzz** [comment]
This example of command might help you https://www.github.com/dubzzz/fast-check/tree/master/example%2F004-stateMachine%2FmusicPlayer%2Fmodel-based%2FAddTrackCommand.ts 🤔

**Comment by matheus23** [comment]
Having this exact issue. I'm working around this by calling `.chain` in a loop, basically implementing the above mentioned `unfold` using `.chain` myself, except that I hard-code the amount of calls my loop should generate (you could generate and `.chain` that as well, but that just doesn't shrink well anyway).

It has worked fine for me so far, but I wonder whether something based on `unfold` might have more information and be able to shrink better than a loop of `.chain`s.

To add a little context around all of this: I want to mostly test *valid* inputs of my reducer. To generate a valid action, I need to know the current state of the reducer, because actions often refer to the current state via IDs, indices, etc.

I think there's going to be many more like me who're trying to test their reducers like that. In my case, I'm essentially writing a virtual filesystem and want to test a series of different commands on it (e.g. write to file X, remove file X, mkdir X, move file X to Y, etc.). To actually test code and not end up in the "this path doesn't exist" case all of the time, I need to know the current filesystem state.

Given `.chain` *exists* (and is basically a monad bind operation and thus allows implementing basically anything), but it's *discouraged* due to shrinking not being ideal, does it make sense to do something similar for `scenario` or `unfold`? I.e. implement it, but document that you'd rather use the commands API or you'll have less ideal shrinking?
I still think there's better algorithms for shrinking `scenario` than `chain`s (which people will then inevitably use instead).

**Comment by dubzzz** [comment]
@matheus23 I initially POC-ed a `chain`-based solution a while ago but it suffered from poor shrinking capacities (see https://github.com/dubzzz/fast-check/pull/445). At that time, fast-check was not powerful enough to be able to shrink 'user definable' values with built-in generators thus it would have required to implement yet another shrinker for arrays.

With the recent rewrite of arbitraries to implement the API NextArbitrary instead of Arbitrary, the arbitrary of arrays is now able to shrink any array.

**Why is this useful for the scenario feature?** Basically the `chain`-based can easily generate an array of valid commands but at the end it has not a very efficient shrink. On the contrary, the arbitrary of arrays can take one of the generated arrays and suggest shrinks for it. It's now possible to combine both and benefit from the strength of both of them.

**Comment by dubzzz** [comment]
In order to leverage built-in shrinkers within custom arbitraries having poor shrinking capacities you can refer to my example: https://github.com/dubzzz/fast-check/issues/1503#issuecomment-851272804

It will be part of the official documentation as soon as I release v3 (no more need to fromNext/toNext in v3).
